File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const threeSum = nums => {
    nums.sort((a, b) => a - b);
    const ans = [];
    for (let i = 0; i < nums.length; i++) {
        const pivot = nums[i];
        let left = i + 1, right = nums.length - 1;
        let sum = pivot;
        if (pivot === nums[i - 1]) {
            continue;
        } else {
            while (left < right) {
                if (sum + nums[left] + nums[right] === 0) {
                    ans.push([pivot, nums[left], nums[right]]);
                    left++;
                    right--;
                    while (left < right && nums[left] === nums[left - 1]) left++;
                } else if (sum + nums[left] + nums[right] < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }
    return ans;
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        pivot = 0
        left = 1
        right = len(nums) - 1
        while pivot < len(nums):
            totalSum = nums[pivot]
            if totalSum > 0:
                break
            while left < right:
                if totalSum + nums[left] + nums[right] == 0 and [nums[pivot], nums[left], nums[right]] not in res:
                    res.append([nums[pivot], nums[left], nums[right]])
                    left += 1
                    right -= 1
                elif totalSum + nums[left] + nums[right] > 0:
                    right -= 1
                else:
                    left += 1
            pivot += 1
            left = pivot + 1
            right = len(nums) - 1
        return res